src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixReorg.java

```diff
@@ -389,10 +393,78 @@ public static MatrixBlock rev( MatrixBlock in, MatrixBlock out ) {
 		return out;
 	}
 
+	public static MatrixBlock rev(MatrixBlock in, MatrixBlock out, int k) {
+		if (k <= 1 || in.isEmptyBlock(false)) {
+			System.out.println("choosing single thread");
+			return rev(in, out); // fallback to single-threaded
+
+		}
+		final int numRows = in.getNumRows();
+		final int numCols = in.getNumColumns();
+		final boolean sparse = in.isInSparseFormat();
+
+		// Prepare output block
+		out.reset(numRows, numCols, sparse);
+
+		// Before starting threads, ensure the output sparse block is allocated!
+		if (sparse) {
+			out.allocateSparseRowsBlock(false);
+		}
+
+		// Set up thread pool
+		ExecutorService pool = CommonThreadPool.get(k);
+		try {
+			int blklen = (int) Math.ceil((double) numRows / k);
+			List<Future<?>> tasks = new ArrayList<>();
+
+			for (int i = 0; i < k; i++) {
+				final int startRow = i * blklen;
+				final int endRow = Math.min((i + 1) * blklen, numRows);
+
+				tasks.add(pool.submit(() -> {
+					if (!sparse) {
```

Comment:
create a static method for this kernel which is called from both the single-threaded implementation as well as the multi-threaded implementation